File: src/Web/TypeScriptTS/example1.ts
// Abstract Product
interface Button {
    render(): void;
}

interface Checkbox {
    render(): void;
}

// Concrete Product
class DarkButton implements Button {
    render(): void {
        console.log("Dark Button");
    }
}

class LightButton implements Button {
    render(): void {
        console.log("Light Button");
    }
}

class DarkCheckbox implements Checkbox {
    render(): void {
        console.log("Dark Checkbox");
    }
}

class LightCheckbox implements Checkbox {
    render(): void {
        console.log("Light Checkbox");
    }
}

// Abstract Factory
interface UIFactory {
    createButton(): Button;
    createCheckbox(): Checkbox;
}

// Concrete Factory
class DarkFactory implements UIFactory {
    createButton(): Button {
        return new DarkButton();
    }
    
    createCheckbox(): Checkbox {
        return new DarkCheckbox();
    }
}

class LightFactory implements UIFactory {
    createButton(): Button {
        return new LightButton();
    }
    
    createCheckbox(): Checkbox {
        return new LightCheckbox();
    }
}

function createUIComponents(factory: UIFactory): void {
    const button = factory.createButton();
    const checkbox = factory.createCheckbox();
    button.render();
    checkbox.render();
}

// Usage
createUIComponents(new DarkFactory());
createUIComponents(new LightFactory());

File: src/Web/Dart/example1.dart
// Abstract Product
abstract class Button {
  void render();
}

abstract class Checkbox {
  void render();
}

// Concrete Product
class DarkButton implements Button {
  @override
  void render() {
    print('Dark Button');
  }
}

class LightButton implements Button {
  @override
  void render() {
    print('Light Button');
  }
}

class DarkCheckbox implements Checkbox {
  @override
  void render() {
    print('Dark Checkbox');
  }
}

class LightCheckbox implements Checkbox {
  @override
  void render() {
    print('Light Checkbox');
  }
}

// Abstract Factory
abstract class UIFactory {
  Button createButton();
  Checkbox createCheckbox();
}

// Concrete Factory
class DarkFactory implements UIFactory {
  @override
  Button createButton() {
    return DarkButton();
  }

  @override
  Checkbox createCheckbox() {
    return DarkCheckbox();
  }
}

class LightFactory implements UIFactory {
  @override
  Button createButton() {
    return LightButton();
  }

  @override
  Checkbox createCheckbox() {
    return LightCheckbox();
  }
}

void createUIComponents(UIFactory factory) {
  var button = factory.createButton();
  var checkbox = factory.createCheckbox();
  button.render();
  checkbox.render();
}

// Usage
void main() {
  createUIComponents(DarkFactory());
  createUIComponents(LightFactory());
}
